ghost/core/core/server/services/Users.js

```diff
@@ -76,68 +77,72 @@ class Users {
     }
 
     async assignTagToUserPosts({id, context, transacting}) {
-        // create an internal tag to assign to reassigned posts
-        // in following format: `#{author_slug}`
+        let createdTag = false;
+        // get author slug
         const author = await this.models.User.findOne({
             id
         }, {
             id,
             context,
             transacting
         });
+        // create an internal tag to assign to reassigned posts
+        // in following format: `#{author_slug}`
         let tag = await this.models.Tag.findOne({
             slug: `hash-${author.get('slug')}`
         }, {
             context,
             transacting
         });
-
         if (!tag) {
             tag = await this.models.Tag.add({
                 slug: `#${author.get('slug')}`
             }, {
                 context,
                 transacting
             });
+            createdTag = true;
         }
 
+        // get list of posts that need the tag assigned
         const userPosts = await this.models.Base.knex('posts_authors')
             .transacting(transacting)
             .where('author_id', id)
             .select('post_id');
-        const usersPostIds = userPosts.map(p => p.post_id);
-
-        // Add a tag to all posts that do not have the author tag yet
-        // NOTE: the method is implemented in an iterative way to avoid
-        //       memory consumption in case the user has thousands of posts
-        //       assigned to them. Also, didn't have any "bulk" way to add
-        //       a tag to multiple posts as the "sort_order" needs custom
-        //       logic to be run for each post.
-        //       Rewrite this bit if/when we hit a performance bottleneck here
-        for (const postId of usersPostIds) {
-            const post = await this.models.Post.findOne({
-                id: postId,
-                status: 'all'
-            }, {
-                id: postId,
-                withRelated: ['tags'],
-                context,
-                transacting
-            });
-
-            // check if tag already assigned to the post
-            const existingTagSlugs = post.relations.tags.models.map(t => t.get('slug'));
 
-            if (!existingTagSlugs.includes(tag.get('slug'))) {
-                await this.models.Post.edit({
-                    tags: [...post.relations.tags.models, tag]
-                }, {
-                    id: postId,
-                    context,
-                    transacting
-                });
-            }
+        // get the posts that do not have the tag already assigned
+        let usersPostIds = [];
```

Comment:
v. minor, style only

It seems to me that `createdTag` is almost always gonna be true - it's like an edge case if it's not?

If that's the case, I would prefer that we make it a lil more explicit like

```js
let usersPostIds = userPosts.map(p => p.post_id);

// CASE: tag already existing and we don't wanna add it twice.
if (!createdTag) {
    const tagId = tag.get('id');
    const taggedPostIds = await this.models.Base.knex('posts_tags')
        .transacting(transacting).where('tag_id', tagId).select('post_id');

    userPostIds = userPostIds.filter(p => !taggedPostIds.includes(p));
}

Reply:
Ha yeah I thought about this as I was writing it and meant to come back to it. Agreed - it'll pretty much always be the case. Almost so much so that we shouldn't bother handling it. (almost)

Reply:
(almost) 😝